File: 2023/python/day02.py
#!/usr/bin/env python3

import sys

p1 = p2 = 0

for game, line in enumerate(sys.stdin.readlines()):
  rounds = [
    {p[1]: int(p[0]) for p in [c.split() for c in r.split(",")]}
    for r in line.strip().split(":")[1].split(";")
  ]

  r = max([x.get("red", 0) for x in rounds])
  g = max([x.get("green", 0) for x in rounds])
  b = max([x.get("blue", 0) for x in rounds])

  if r <= 12 and g <= 13 and b <= 14:
    p1 += game + 1

  p2 += r * g * b

print("02:", p1, p2)

File: 2023/kotlin/src/Day02.kt
import java.io.File

fun main(args: Array<String>) {
    val games = File(args.getOrNull(0) ?:
            "../../inputs/2023/02-input.txt")
        .readLines()
        .map { line -> line
            .split(':')[1].trim()
            .split(';')
            .map { round ->
                round.split(',')
                    .map { it.trim().split(' ') }
                    .associateBy( { it[1] }, { it[0].toInt() })
            }
        }

    val p1 = games.withIndex()
        .filter { (_, rounds) ->
            rounds.all {
                it.getOrDefault("red", 0) <= 12 &&
                it.getOrDefault("green", 0) <= 13 &&
                it.getOrDefault("blue", 0) <= 14
            }
        }
        .sumOf { it.index + 1 }

    val p2 = games
        .sumOf { rounds ->
            rounds.maxOf { it.getOrDefault("red", 0) } *
            rounds.maxOf { it.getOrDefault("green", 0) } *
            rounds.maxOf { it.getOrDefault("blue", 0) }
        }

    print("02: $p1 $p2")
}
